### PR TITLE
Remove MemcpyOptPass from O2 for Enzyme

### PR DESCRIPTION
**Pull request by rmoyard**
**Context:**
Currently test are failing with Enzyme because it does not handle well `memset` and `memcpy` operations, some of them are generated by O2.

**Description of the Change:**
We remove the `MemcpyOptPass` from the pass manager using the pass instrumentation.

**Benefits:**
More Enzyme bugs fixed.

### CODE REVIEW

**Comment by rmoyard** [comment]
[sc-66104]

**Comment by rmoyard** [comment]
Fix the wrong results in [sc-59758]

File: mlir/lib/Driver/CompilerDriver.cpp [hunk 1]
```diff
@@ -443,7 +443,14 @@ LogicalResult runO2LLVMPasses(const CompilerOptions &options,
...
     // options.
-    llvm::PassBuilder PB;
+    llvm::PassInstrumentationCallbacks PIC;
+    PIC.registerShouldRunOptionalPassCallback([](llvm::StringRef P, llvm::Any) {
+        if (P == "MemCpyOptPass") {
+            return false;
+        }
+        return true;
+    });
+    auto PB = llvm::PassBuilder(nullptr, llvm::PipelineTuningOptions(), std::nullopt, &PIC);
```
Comment (dime10) [inline mlir/lib/Driver/CompilerDriver.cpp]:
```suggestion
    llvm::PassBuilder PB(nullptr, llvm::PipelineTuningOptions(), std::nullopt, &PIC);
```

**Review by dime10 — APPROVED** [review]
Nice fix :) Let's merge this into the rc

**Pull request merged**